Add self-join and self-leave commands

calebfletcher opened the pull request:
Rename ROLE_NAME config variable to differentiate it from the self-join role names

SELF_JOIN_ALIASES allows commands to be created that alias the self-join command, e.g. `!notify` to `!join notified`. The config variable is an object with each key as an alias command and the associated value the full command.

calebfletcher committed:
Add self-join command
Rename ROLE_NAME config variable to differentiate it from the self-join 
role names

seanbudd commented on config.json.template:
```diff
@@ -3,13 +3,19 @@
+    "CRYPTO_JSON_MEMBER_ENCRYPT_KEY": "",
+    "SELF_JOIN_COMMAND": "!join",
+    "SELF_JOIN_REQUIRE_VERIFIED": true,
+    "SELF_JOIN_ROLE_NAMES": [
```
can this become a comma separated string? 
Environment variables with nodes process.env system only allows for strings for key/values

calebfletcher replied:
Latest commit now makes this config variable a comma separated string, with any spaces on either side of a comma being stripped automatically

seanbudd commented on config.json.template:
```diff
@@ -3,13 +3,19 @@
+    "SELF_JOIN_REQUIRE_VERIFIED": true,
+    "SELF_JOIN_ROLE_NAMES": [
+    ],
+    "SELF_JOIN_ALIASES": {
```
this too needs to be a string. 

Not sure how we can support aliases :( 

First idea is to just make it a json dump in a string that can be parsed on load?

seanbudd replied:
I think the easiest option is to strip it entirely :P KISS?

calebfletcher replied:
Removed it for now, maybe something to do for a future feature

seanbudd commented on index.js:
```diff
@@ -50,6 +51,46 @@ client.on('message', message => {
         .send(MESSAGE_PREFIX + "Send '!verify' to access other channels")
         .catch(reason => console.log(reason))
     }
+    else if (text.startsWith(CONFIG.SELF_JOIN_COMMAND) || is_self_join_alias) {
```
can we do regex matching here? If you're new to regex I'm happy to do the change :)

calebfletcher replied:
Latest commit now parses this with regex

seanbudd commented on index.js:
```diff
@@ -50,6 +51,46 @@ client.on('message', message => {
+    else if (text.startsWith(CONFIG.SELF_JOIN_COMMAND) || is_self_join_alias) {
+      let guild = client.guilds.get(CONFIG.GUILD_ID)
+      let msg = is_self_join_alias ? CONFIG.SELF_JOIN_ALIASES[text] : text
+      let msg_roles = msg.replace('!join', '').trim()
```
regex match would automatically parse the role ;)

calebfletcher replied:
Latest commit also parses this through regex

seanbudd commented on index.js:
```diff
@@ -50,6 +51,46 @@ client.on('message', message => {
+                    if (member.roles.has(guild_role.id)) {
+                      member
+                        .removeRole(guild_role)
+                        .then(message.channel.send(`Left role '${msg_role}'`).catch(reason => console.log(reason)))
```
I don't like the idea of leaving a role with !join :/ 

Would it be too much work to make !leave for this purpose?

seanbudd replied:
and let people know they are already in the role instead

calebfletcher replied:
Added a separate command for leaving, with appropriate logic to notify the user if they are already in the role and are trying to join, or if they are not in the role and trying to leave

seanbudd reviewed (changes requested):
Hey Caleb,

Thanks for the PR, I've left a few comments of suggested changes needed for compatibility

Let me know if you need any help

Cheers,
Sean

calebfletcher committed:
Fix compatibility issues

calebfletcher committed:
Update config template

calebfletcher commented:
Updated most of the joining and leaving logic, breaking it out into two separate commands, both configurable. The config variable SELF_JOIN_ROLE_NAMES also now takes a comma-delimited string of role names, rather than a JSON array. Regex has also replaced most of the trimming and string parsing methods.

seanbudd commented on config.json.template:
```diff
@@ -3,13 +3,17 @@
     "CRYPTO_JSON_ENCODING": "hex",
-    "CRYPTO_JSON_MEMBER_ENCRYPT_KEY": ""
+    "CRYPTO_JSON_MEMBER_ENCRYPT_KEY": "",
+    "SELF_JOIN_JOIN_COMMAND": "",
```
```suggestion
    "SELF_JOIN_JOIN_COMMAND": "!join",
```

seanbudd commented on config.json.template:
```diff
@@ -3,13 +3,17 @@
-    "CRYPTO_JSON_MEMBER_ENCRYPT_KEY": ""
+    "CRYPTO_JSON_MEMBER_ENCRYPT_KEY": "",
+    "SELF_JOIN_JOIN_COMMAND": "",
+    "SELF_JOIN_LEAVE_COMMAND": "",
```
```suggestion
    "SELF_JOIN_LEAVE_COMMAND": "!leave",
```

I did like the ol' defaults B)

seanbudd commented on index.js:
```diff
@@ -50,6 +50,57 @@ client.on('message', message => {
         .send(MESSAGE_PREFIX + "Send '!verify' to access other channels")
         .catch(reason => console.log(reason))
     }
+    else if (new RegExp(`^${CONFIG.SELF_JOIN_JOIN_COMMAND}|${CONFIG.SELF_JOIN_LEAVE_COMMAND}`).test(text)) {
```
```suggestion
    else if (new RegExp(`^${CONFIG.SELF_JOIN_JOIN_COMMAND}`).test(text)) {
```

and can the leave stuff be a separate, top-level condition?

The amount nested if conditions makes this difficult to work with :(

seanbudd replied:
can we**

amount of**

seanbudd commented on index.js:
```diff
@@ -50,6 +50,57 @@ client.on('message', message => {
+                  message.channel.send(`Cannot find role '${role_name}'`).catch(reason => console.log(reason))
+                }
+              }
+            } while ((arrayRoles = regexRoles.exec(text)) !== null)
```
nonnonono please no while loop, especially if this can be done with a for loop

side note: is it really necessary to have an array of roles? simplest (perhaps not best) UI is to just join them individually with multiple usages

calebfletcher replied:
Removed the ability to join/leave multiple roles in a single command, it now accepts anything after the command as a single role name

seanbudd reviewed (commented):
Hey @calebfletcher,

Thanks for incorporating the previous suggestions. I've left some more feedback with the goal of moving away from the 9ish levels of nested blocks, as it makes the code harder to maintain. 

Thanks,
Sean

seanbudd committed:
Apply suggestions from code review

Co-Authored-By: calebfletcher <caleb@fletcher.cf>

calebfletcher committed:
Split join and leave commands into seperate logic blocks
Make join and leave commands accept only one role name

seanbudd commented on index.js:
```diff
@@ -50,6 +50,80 @@ client.on('message', message => {
+          }
+        })
+        .catch(reason => console.log(reason))
+    }
```
I've rewritten the codeblock for !join to be a bit clearer, could you implement something like this?

```javascript

    } else if (text.match(/^!join ([^!\s]*)\s*$/)) {
      const verified_role = message.guild.roles.find(role => role.name === CONFIG.VERIFIED_ROLE_NAME)
      const requested_role_name = text.match(/^!join ([^!\s]*)\s*$/)[0]
      if (CONFIG.SELF_JOIN_REQUIRE_VERIFIED && !message.member.roles.has(verified_role.id)) {
        message.channel.send('Please verify before trying to join a role').catch(reason => console.log(reason))
        return
      }
      if (!requested_role_name) {
        message.channel.send('Please specify the role you would like to join').catch(reason => console.log(reason))
        return
      }
      const requested_role = message.guild.roles.find(role => role.name === requested_role_name)
      if (!requested_role) {
        message.channel.send(`Cannot find role '${requested_role_name}'`).catch(reason => console.log(reason))
        return
      }
      if (!CONFIG.SELF_JOIN_ROLE_NAMES.split(/\b\s*,\s*\b/g).includes(requested_role_name)) {
        message.channel
          .send(`You are not allowed to join '${requested_role_name}'`)
          .catch(reason => console.log(reason))
        return
      }
      if (!message.member.roles.has(requested_role.id)) {
        message.channel
          .send(`You are already in role '${requested_role_name}', try '!leave ${requested_role_name}' to leave`)
          .catch(reason => console.log(reason))
        return
      }
      message.member
        .addRole(requested_role)
        .then(message.channel.send(`Joined role '${requested_role_name}'`).catch(reason => console.log(reason)))
```

seanbudd commented:
Hey @calebfletcher I hate to ruin your hard work but I think we need to expand this feature beyond what was originally intended. See #3 for more information
